File: 9_排序和搜索/力扣加加/插入排序二分/327. 区间和的个数.py
#  给你一个整数数组 nums 以及两个整数 lower 和 upper 。
#  求数组中，值位于范围 [lower, upper] （包含 lower 和 upper）之内的 区间和的个数 。
#  区间和sum(i,j) = S[j] - S[i]

import bisect
from typing import List


def solution(list: List[int], low: int, upper: int):
    res = 0
    sum: List[int] = [0]
    cur_sum = 0
    for value in list:
        cur_sum += value
        res += bisect.bisect_right(sum, cur_sum - low) - bisect.bisect_left(sum, cur_sum - upper)
        sum.append(cur_sum)
    return res


print(solution([-2, 5, -1], -2, 2))

File: 22_专题/区间和/前缀/327. 区间和的个数.ts
// 如果数组是非负的，那么前缀和就是一个单调不递减数组，我们有时候可以基于它来做二分。
// 数组 A 有多少个连续的子数组，其元素只和在 [lower, upper]的范围内。
// 1 <= nums.length <= 10**5
// 思路：遍历数组中的每一个元素，求得到当前元素的前缀和

import { bisectLeft } from '../../../9_排序和搜索/二分api/7_二分搜索寻找最左插入位置'
import { bisectRight } from '../../../9_排序和搜索/二分api/7_二分搜索寻找最插右入位置'
import { bisectInsort } from '../../../9_排序和搜索/二分api/7_二分搜索插入元素'

// lower<=Si-Sj<=upper 等价于 Si-upper<=Sj<=Si-lower
function countRangeSum(nums: number[], lower: number, upper: number): number {
  const pre = [0]
  let res = 0
  let sum = 0

  for (const num of nums) {
    sum += num
    res += bisectRight(pre, sum - lower) - bisectLeft(pre, sum - upper)
    bisectInsort(pre, sum)
  }

  return res
}

console.log(countRangeSum([-2, 5, -1], -2, 2))

export {}
